Optimize dispatch for non `__torch_function__` calls

robieta opened the pull request:
This PR adds several optimizations:
1) ~~Define `torch._C._all_tensors` method which is much faster than `all(type(t) is Tensor for t in tensors)` (50 ns vs. 250 ns / element), and roll the fast path bailout into `has_torch_function`~~

2) ~~Define `torch._C._is_tensor` method to expose `THPVariable_CheckExact` in Python. This is marginally faster than `type(t) is Tensor` (50 ns vs 85 ns), and it's worth keeping `not _is_tensor(t) and has_torch_function((t,))` because it lets us skip a function call and tuple construction, which works out to about 50 ns savings over just `has_torch_function`. (And there are a lot of unary functions.)~~

3) ~~Reorder the checks in `torch.functional` and `torch.nn.functional`, as in most cases `torch.jit.is_scripting()` is False (and we are more overhead sensitive in eager), so for the eager fast path we can save >100 ns by checking that after.~~

4) Stop importing `torch.overrides` inside functions now that the circular import is gone.

**EDIT: See below for updated changes.**

This PR is pretty big, but also mostly mechanical changes.

The test script assumes two envs: `ref` which is built from master, and `fast_torch_fn_check` which is built from this PR. It's really hard to get stable overhead measurements, so I went for the slightly more complicated but much more precise "throw compute at it" approach. This PR covers several dozen ops, so I selected one of each code path (e.g. `has_torch_function(tensors)`, `not _is_tensor(input) and has_torch_function((input,))`, etc.) and annotated them appropriately.

<details>

<summary> test.py </summary>

```
import argparse
import multiprocessing
import multiprocessing.dummy
import os
import pickle
import queue
import random
import sys
import subprocess
import tempfile
import time

import torch
from torch.utils.benchmark import Timer, Compare


NUM_CORES = multiprocessing.cpu_count()
ENVS = (
    "ref",
    "fast_torch_fn_check"
)


MIN_RUN_TIME = 1
REPLICATES = 300
SETUP = """
x = torch.ones((1, 1))
y = torch.ones((1, 1))
"""

TASKS = {
    "tensor.py: _wrap_type_error_to_not_implemented `__floordiv__`": "x // y",
    "tensor.py: method          `__hash__`": "hash(x)",
    "functional.py: (unary)     `unique`": "torch.functional.unique(x)",
    "functional.py: (args)      `atleast_1d`": "torch.functional.atleast_1d((x, y))",
    "nn/functional.py: (unary)  `relu`": "torch.nn.functional.relu(x)",
    "nn/functional.py: (args)   `linear`": "torch.nn.functional.linear(x, y)",
}


def worker_main(argv):
    parser = argparse.ArgumentParser()
    parser.add_argument("--output_file", type=str)
    output_file = parser.parse_args(argv).output_file

    env = os.path.split(os.getenv("CONDA_PREFIX"))[1]
    assert env in ENVS

    results = []
    for i, stmt in enumerate(TASKS.values()):
        timer = Timer(
            stmt=stmt,
            setup=SETUP,
            env=env,
            sub_label=" ",
            description=f"[{i}]",
        )
        results.append(timer.blocked_autorange(min_run_time=MIN_RUN_TIME))

    with open(output_file, "wb") as f:
        pickle.dump(results, f)


def main():
    num_workers = int(NUM_CORES // 2)
    tasks = list(ENVS * REPLICATES)
    random.shuffle(tasks)
    task_queue = queue.Queue()
    for t in tasks:
        task_queue.put(t)
    results = []

    def map_fn(worker_id):
        core = str(worker_id * 2)
        _, output_file = tempfile.mkstemp(suffix=".pkl")
        try:
            while True:
                try:
                    env = task_queue.get_nowait()
                except queue.Empty:
                    break

                subprocess.run(
                    " ".join([
                        "source", "activate", env, "&&",
                        "taskset", "--cpu-list", core,
                        "python", os.path.abspath(__file__),
                        "--mode", "worker",
                        "--output_file", output_file
                    ]),
                    shell=True,
                    check=True,
                )

                # We don't need a lock, as the GIL is enough.
                with open(output_file, "rb") as f:
                    results.extend(pickle.load(f))
        finally:
            os.remove(output_file)

    with multiprocessing.dummy.Pool(num_workers) as pool:
        st, eta, n_total = time.time(), "", len(tasks) * len(TASKS)
        map_job = pool.map_async(map_fn, range(num_workers))
        while not map_job.ready():
            n_complete = len(results)
            if n_complete:
                sec_per_element = (time.time() - st) / n_complete
                n_remaining = n_total - n_complete
                eta = f"ETA: {n_remaining * sec_per_element:.0f} sec"

            print(f"\r{n_complete} / {n_total}   {eta}".ljust(40), end="")
            sys.stdout.flush()
            time.sleep(2)
        print()

    compare = Compare(results)
    compare.trim_significant_figures()
    compare.colorize()
    compare.print()
    for i, k in enumerate(TASKS.keys()):
        print(f"[{i}] {k}")
    print()


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--mode", type=str, choices=("main", "worker"), default="main")
    args, remaining = parser.parse_known_args()

    if args.mode == "main":
        assert not remaining
        main()
    else:
        worker_main(remaining)

```

</details>

<img width="676" alt="Screen Shot 2020-12-03 at 12 04 09 AM" src="https://user-images.githubusercontent.com/13089297/100981059-1cd3a480-34fb-11eb-89c1-78d73f7700f2.png">

CC @ezyang @bhosmer @xwang233 @mcarilli @mruberry

Commit:
reduce overhead from checking for __torch_function__

Commit:
move  check after torch_fn check

hameerabbasi commented on torch/functional.py:
```diff
@@ -200,12 +199,11 @@ def lu_unpack(LU_data, LU_pivots, unpack_data=True, unpack_pivots=True):
         >>> torch.norm(A_ - A)
         tensor(2.9802e-08)
     """
-    if not torch.jit.is_scripting():
```
I wasn't sure the tests would fail but they did...

The torch JIT isn't capable of `__torch_function__`, so it bails on things in here. Writing the `if` statement like this makes it so the JIT can handle things properly.

ezyang replied:
@suo how would you feel if we add a special case to TorchScript to identify blocks like this and treat them specially in the same way as `is_scripting()`? We can change the precise wording of the condition here as necessary, but with the constraint that eager mode perf has to be as fast as possible.

suo replied:
I think that's reasonable, given that this block is hit from everything ever. We can define a `_has_torch_function_noscript` or something equally unlikely for users to use and have that is statically resolved to false in TS. cc @gmagogsfm @SplitInfinity, what do you think?

SplitInfinity replied:
Is the goal here to merge the two predicates of these if statements into one?

ezyang replied:
The goal here is to reduce eager overhead. Merging all of these predicates together into a single predicate would certainly work, although we still have to be careful not to accidentally start allocating tuples to make this call.

robieta replied:
Tests are failing because the schema of `has_torch_function` is not valid. (It should be `Any[]`, but that's not a thing.) Some initial grep-ing indicates that the solution is probably to make a kind of `SugaredValue` which swallows arguments and then register `has_torch_function` as taking no args, but I figured I should check with you folks to see if there is a better way.

robieta replied:
Actually, it looks like `Any` is allowed, so I'm good.

hameerabbasi commented on torch/csrc/Module.cpp:
```diff
@@ -628,6 +645,8 @@ static PyMethodDef TorchMethods[] = {
   {"_is_xnnpack_enabled", THPModule_isEnabledXNNPACK, METH_NOARGS, nullptr},
   {"_is_torch_function_enabled", THPModule_isEnabledTorchFunction, METH_NOARGS, nullptr},
   {"_disabled_torch_function_impl", THPModule_disable_torch_function, METH_VARARGS, nullptr},
+  {"_is_tensor", THPModule_is_tensor, METH_O, nullptr},
```
Can we document that this is an exact `Tensor`, perhaps by changing the name to not confuse people into thinking it's checking for instance as well?

ezyang replied:
Agreed, the underscore makes it better, but people are still going to misunderstand. `_is_tensor_exact` perhaps

Commit:
move has_torch_function check to use Python C API

robieta commented:
I've updated this PR and essentially adopted the approach discussed above. I've moved the logic for `has_torch_function` entirely into C++, and added a special case `object_has_torch_function` which skips tuple creation (and unpacking) for the single argument case since it's so common. And then I simply registered both to the JIT as static `False` methods.

Overall performance looks good: [Updated benchmark script](https://gist.github.com/robieta/3bf55fcaf23c0be35f10e8599aa6d87f)
<img width="1166" alt="Screen Shot 2020-12-06 at 3 44 45 PM" src="https://user-images.githubusercontent.com/13089297/101296745-fb640880-37d9-11eb-9d40-c7ab757fcaf4.png">

(Added before and after https://github.com/pytorch/pytorch/pull/37091 as well.) I'm still not thrilled with `Module` performance. The large difference in overhead for the later rows is due to passing `Parameter` rather than `Tensor` into `matmul`, so something is wonky on the binding front. (Maybe the Python arg parser.) Outside the scope of this PR, but definitely due for a follow up PR.

Commit:
Merge branch 'master' into gh/taylorrobie/torch_fn_overhead

robieta commented:
Also, I just realized I didn't add entries in the `.pyi` file. I'll do that when I fix the lint errors.

robieta commented:
So apparently we are codegen-ing:

```
if(check_has_torch_function(self_))
```

rather than
```
if(!THPVariable_CheckExact(self_) && check_has_torch_function(self_))
```

(or `THPVariableOrParameter_CheckExact` / whatever we wind up calling the version that treats Parameters as Tensors) for no-arg ops (e.g. `x.t()`), so `Tensor` is going through the slower `check_has_torch_function`, and `Parameter` calls all the way back to Python.

For `.t()` it's about 0.3% instruction delta (50 ns, 2% wall time) on Tensor and 13% instructions (17% wall time) on Parameter.

hameerabbasi commented:
> and `Parameter` calls all the way back to Python.

Could you clarify this? Technically, we check for exact `Tensor`, and we also check for `type.__torch_function__ is _disabled_torch_function_impl` attribute.

robieta commented:
> Could you clarify this? Technically, we check for exact `Tensor`, and we also check for `type.__torch_function__` is `_disabled_torch_function_impl` attribute.

**python_variable_methods.cpp**
```
if(check_has_torch_function(self_)) {
  return handle_torch_function(self_, "t");
}
```

`check_has_torch_function` calls `PyTorch_LookupSpecial` which does an exact check, but allocates an extra `py::object` and does some pointer checks in the process. And `Parameter` goes through the last `PyObject_FastGetAttrString` path.

hameerabbasi commented:
Apologies, pressed the wrong button.

`py::object` is a `PyObject*` essentially, so return `py::object()` is `return nullptr`.

robieta commented:
> Apologies, pressed the wrong button.
> 
> `py::object` is a `PyObject*` essentially, so return `py::object()` is `return nullptr`.

I believe `py::handle` is the cheap one, and `py::object` also does ref counts. (It may be smart enough not to do it for the null case.) It's not a huge cost for Tensor, but it is slower than doing a type check at the start.

ezyang commented:
> For .t() it's about 0.3% instruction delta (50 ns, 2% wall time) on Tensor and 13% instructions (450 ns, 17% wall time) on Parameter.

Maybe this is the smoking gun for the other CUDA problem?

robieta commented:
> > For .t() it's about 0.3% instruction delta (50 ns, 2% wall time) on Tensor and 13% instructions (450 ns, 17% wall time) on Parameter.
> 
> Maybe this is the smoking gun for the other CUDA problem?

Possible, though I'm skeptical. I only noticed it because `Linear` calls `input.matmul(weight.t())`; however the CUDA issue is a ResNet and I don't see anything similar on conv2d, bn, or relu.

robieta commented:
By the way, this PR is getting quite big and complicated (it was initially just an exploratory PR), so I'm going to break it into a stack.
